File: java/Trees/RightmostNodesOfABinaryTree.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

import DS.TreeNode;

/*
    // Definition of TreeNode:
    class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int val) {
            this.val = val;
        }
    }
 */

public class RightmostNodesOfABinaryTree {
    public List<Integer> rightmostNodesOfABinaryTree(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<Integer> res = new ArrayList<>();
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            // Add all the non-null child nodes of the current level to the 
            // queue.
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
                // Record this level's last node to the result array.
                if (i == levelSize - 1) {
                    res.add(node.val);
                }
            }
        }
        return res;
    }
}

File: kotlin/Trees/RightmostNodesOfABinaryTree.kt
import java.util.LinkedList
import ds.TreeNode

/*
    Definition of TreeNode:

    data class TreeNode(
        var value: Int, 
        var left: TreeNode? = null, 
        var right: TreeNode? = null
    )
*/

fun rightmostNodesOfABinaryTree(root: TreeNode?): List<Int> {
    val res = mutableListOf<Int>()
    if (root == null) {
        return res
    }
    val queue = LinkedList<TreeNode>()
    queue.add(root)
    while (queue.isNotEmpty()) {
        val levelSize = queue.size
        // Add all the non-null child nodes of the current level to the
        // queue.
        repeat(levelSize) {
            val node = queue.poll()
            if (node.left != null) {
                queue.add(node.left!!)
            }
            if (node.right != null) {
                queue.add(node.right!!)
            }
            // Record this level's last node to the result array.
            if (it == levelSize - 1) {
                res.add(node.value)
            }
        }
    }
    return res
}
